tsl/src/remote/connection.c

```diff
@@ -697,6 +697,16 @@ remote_connection_configure(TSConnection *conn)
 		i++;
 	}
 
+	/*
+	 * If we are initiating connection from a standby (of an AN for example),
+	 * then the remote connection session needs to be also set up as a
+	 * READ ONLY one. This will catch any commands that are sent from the
+	 * read only AN to datanodes but which could have potential read-write
+	 * side effects on data nodes.
+	 */
+	if (RecoveryInProgress())
```

Comment:
Shouldn't we also check for other read-only situations, not only recovery-in-progress. For instance, the session on the AN could also have done `SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY`, or  `default_transaction_read_only=true` is set.

Reply:
Is it enough to check `XactReadOnly`?

Reply:
> Is it enough to check `XactReadOnly`?

On AN, `XactReadOnly` can be set for a specific transaction whereas we want to set the entire remote SESSION as read only for the rest of its life. 

Also, we could intercept future `SET SESSION` or `set transaction` kinda calls on the AN backend and do the same on the remote sessions, but then we will also need to trap RESET commands. There is a `remote_connection_configure_if_changed` function which could be modified to handle this.